mozilla/fx-private-relay
# Set up visual regression testing using Playwright

# New feature description

This adds visual regression testing. I've [run it several times in a row](https://github.com/mozilla/fx-private-relay/actions/runs/2004708839), and it does not appear to be flaky - a common problem with visual regression tests. That said, updating the screenshots (in case of intended changes) is a bit cumbersome, because you'll need to trigger a run in CI to make sure the environment (resolution, zoom level, available fonts, etc.) are consistent. So consider this PR a proof-of-concept: it shows that we could do this, but it's a totally valid option to discard it and choose not to go with this. Let me know what you think.

# Screenshot (if applicable)

Well... They're in the diff :grin: 

# How to test

You could try running it locally (don't forget to `npm install`; also `npx playwright install` to download dev versions of the relevant browsers): `npm run ui-test`. The first time it will probably fail, because it's generating the screenshots on your machine. (Assuming you're not on Linux; otherwise it will compare the screenshots to the reference ones, which I'd expect to slightly differ.) Subsequent runs _should_ succeed, however: it will generate the screenshots again, and compare them to the ones it generated before. Now make a visual change to one of the elements in one of the screenshots, run the tests again, and watch them fail, and a test report being generated. (And if this turns out flaky, let me know! We'd probably want to close this if so.)

# Checklist
